python/ray/util/client/server/proxier.py

```diff
@@ -560,61 +568,114 @@ def modify_connection_info_resp(self,
+        if reconnecting:
+            with self.clients_lock:
+                if client_id not in self.clients_last_seen:
+                    # Client took too long to reconnect, session has already
+                    # been cleaned up
+                    context.set_code(grpc.StatusCode.NOT_FOUND)
+                    context.set_details(
+                        "Attempted to reconnect a session that has already "
+                        "been cleaned up")
+                    return
+                self.clients_last_seen[client_id] = start_time
+            server = self.proxy_manager._get_server_for_client(client_id)
+            channel = self.proxy_manager.get_channel(client_id)
+            # iterator doesn't need modification on reconnect
+            new_iter = request_iterator
+        else:
+            # Create Placeholder *before* reading the first request.
+            server = self.proxy_manager.create_specific_server(client_id)
             with self.clients_lock:
+                self.clients_last_seen[client_id] = start_time
                 self.num_clients += 1
 
-            logger.info(f"New data connection from client {client_id}: ")
-            init_req = next(request_iterator)
-            try:
-                modified_init_req, job_config = prepare_runtime_init_req(
-                    init_req)
-                if not self.proxy_manager.start_specific_server(
-                        client_id, job_config):
-                    logger.error(
-                        f"Server startup failed for client: {client_id}, "
-                        f"using JobConfig: {job_config}!")
-                    raise RuntimeError(
-                        "Starting Ray client server failed. See "
-                        f"ray_client_server_{server.port}.err for detailed "
-                        "logs.")
-                channel = self.proxy_manager.get_channel(client_id)
-                if channel is None:
-                    logger.error(f"Channel not found for {client_id}")
-                    raise RuntimeError(
-                        "Proxy failed to Connect to backend! Check "
-                        "`ray_client_server.err` and "
-                        f"`ray_client_server_{server.port}.err` on the head "
-                        "node of the cluster for the relevant logs. "
-                        "By default these are located at "
-                        "/tmp/ray/session_latest/logs.")
-                stub = ray_client_pb2_grpc.RayletDataStreamerStub(channel)
-            except Exception:
-                init_resp = ray_client_pb2.DataResponse(
-                    init=ray_client_pb2.InitResponse(
-                        ok=False, msg=traceback.format_exc()))
-                init_resp.req_id = init_req.req_id
-                yield init_resp
-                return None
-
-            new_iter = chain([modified_init_req], request_iterator)
+        try:
+            if not reconnecting:
```

Comment:
Note for reviewers: The "changes" here are mostly just from the indent that came from adding the `if not reconnecting`. The only real changes are the lines setting the reconnect_grace_period, and the `stub=` being moved farther down so that the reconnecting code path can reuse it